### ISSUE TITLE
`disallowed_methods` on trait impls

### ISSUE THREAD

**Issue by Jarcho** [rust-lang/rust-clippy#8581]
### Description

It should be possible to to specify a specific impl of a trait method (e.g. `<alloc::vec::Vec as core::iter::IntoIterator>::into_iter`) to be disallowed, rather than having to disallow the trait method for all impls. Two things will need to be done to fix this:

* `clippy.toml` syntax will need to be extended. See `clippy_utils::def_path_res`.
* The `DefId` for the trait method will need to be resolved to it's impl's `DefId`, if possible (generic code will need something more creative here as it could be calling `<T as IntoIterator>::into_iter`).

**Comment by safinaskar** [comment]
I agree this would be a good feature. There are many ways to copy a string (`a.to_owned()`, `a.clone()`, `a.to_string()`). This lint will allow us to enforce particular way

**Comment by Emilgardis** [comment]
Is this still being worked on? @DevAccentor

**Comment by emilk** [comment]
One thing I would LOVE to use this for is to forbid using `std::fmt::Debug` for certain types that already implement `Display`.

I'm constantly battling code that uses `Debug` (`{:?}`) to format things that implement `Display` (`{}`) when presenting them to a user (GUI, CLI, error message, …).

**Comment by emilk** [comment]
### Counter-proposal: `disallowed-trait-usage`

Implementing a "disallowed trait method" might be difficult, and not cover all uses.

Say we want to forbid `Debug`-printing of `PathBuf`:

```rs
fn print(x: &dyn Debug) { … }

print(&my_path);
```

In that case, it would be best to have a clippy rule that forbade all of `<std::fmt::Debug for std::path::PathBuf>`.

I think that would work for what OP asked for as well, but instead of forbidding `<alloc::vec::Vec as core::iter::IntoIterator>::into_iter` (ignoring the problem of templates here 😬 ) we would forbid `<alloc::vec::Vec as core::iter::IntoIterator>`

So, I suggest having a `disallowed-trait-usage` lint with:

```toml
disallowed-trait-usage = [
   { type = "std::fs::PathBuf", trait = "std::fmt::Debug", reason = "Use path.display() instead" },
]
```

@Jarcho how does that sound to you?

Should I open a new request, or should we rename this issue?

---

**EDIT:** I've created a separate issue for this:
* https://github.com/rust-lang/rust-clippy/issues/15765

**Comment by asasine** [comment]
I don't think `disallowed_trait_usage` is sufficient, since there may be some methods on a trait that are fine but others that are not.
